File: src/infra/src/cluster_coordinator/alerts.rs
// Copyright 2024 OpenObserve Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use std::{future::Future, sync::Arc};

use bytes::Bytes;
use config::meta::{alerts::alert::Alert, stream::StreamType};
use itertools::Itertools;

use crate::{db::Event, errors::Error};

/// Sends event to the cluster coordinator indicating that an alert has been put
/// into the database.
pub async fn emit_put_event(org: &str, alert: &Alert) -> Result<(), Error> {
    let key = alert_key(org, alert.stream_type, &alert.stream_name, &alert.name);
    let cluster_coordinator = super::get_coordinator().await;
    cluster_coordinator
        .put(&key, bytes::Bytes::from(""), true, None)
        .await?;
    Ok(())
}

/// Sends event to the cluster coordinator indicating that an alert has been
/// deleted from the database.
pub async fn emit_delete_event(
    org: &str,
    stream_type: StreamType,
    stream_name: &str,
    alert_name: &str,
) -> Result<(), Error> {
    let key = alert_key(org, stream_type, stream_name, alert_name);
    let cluster_coordinator = super::get_coordinator().await;
    cluster_coordinator.delete(&key, false, true, None).await
}

/// Watch for alert events published to the cluster coordinator.
pub async fn watch_events<OnPut, OnPutFut, OnDelete, OnDeleteFut>(
    on_put: OnPut,
    on_delete: OnDelete,
) -> Result<(), anyhow::Error>
where
    OnPut: Fn(String, StreamType, String, String, Option<Bytes>) -> OnPutFut,
    OnPutFut: Future<Output = Result<(), anyhow::Error>>,
    OnDelete: Fn(String, StreamType, String, String) -> OnDeleteFut,
    OnDeleteFut: Future<Output = Result<(), anyhow::Error>>,
{
    let cluster_coordinator = super::get_coordinator().await;
    let mut events = cluster_coordinator.watch("/alerts/").await?;
    let events = Arc::get_mut(&mut events).unwrap();
    log::info!("Start watching alerts");
    loop {
        let ev = match events.recv().await {
            Some(ev) => ev,
            None => {
                log::error!("watch_alerts: event channel closed");
                break;
            }
        };
        match ev {
            Event::Put(ev) => {
                let Some((org, stream_type, stream_name, alert_name)) = parse_alert_key(&ev.key)
                else {
                    log::error!("watch_alerts: failed to parse event key {}", &ev.key);
                    continue;
                };
                let _ = (on_put)(org, stream_type, stream_name, alert_name, ev.value).await;
            }
            Event::Delete(ev) => {
                let Some((org, stream_type, stream_name, alert_name)) = parse_alert_key(&ev.key)
                else {
                    log::error!("watch_alerts: failed to parse event key {}", &ev.key);
                    continue;
                };
                let _ = (on_delete)(org, stream_type, stream_name, alert_name).await;
            }
            Event::Empty => {}
        }
    }
    Ok(())
}

/// Returns the key used to identify an individual alert in events sent to
/// cluster cache watchers.
fn alert_key(org: &str, stream_type: StreamType, stream_name: &str, alert_name: &str) -> String {
    format!("/alerts/{org}/{stream_type}/{stream_name}/{alert_name}")
}

/// Tries to parse the key used to identify an individual alert in avents
/// sent to cluster cache watchers. Returns the organization, stream type,
/// stream name, and alert name from the key.
fn parse_alert_key(key: &str) -> Option<(String, StreamType, String, String)> {
    let parts = key.trim_start_matches("/").split('/').collect_vec();
    if parts.len() < 5 || parts[0] != "alerts" {
        return None;
    }
    let org = parts[1].to_owned();
    let stream_type: StreamType = parts[2].into();
    let stream_name = parts[3].to_owned();
    let alert_name = parts[4].to_owned();
    Some((org, stream_type, stream_name, alert_name))
}

File: web/src/test/unit/mockData/alerts.ts
// Copyright 2023 OpenObserve Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

export default {
  alerts: {
    get: {
      list: [
        {
          name: "alert1",
          stream: "default",
          query: {
            sql: "select * from default",
            from: 0,
            size: -1,
            start_time: 0,
            end_time: 0,
            query_type: "logs",
            track_total_hits: false,
          },
          condition: {
            column: "code",
            operator: "EqualTo",
            ignore_case: null,
            value: 500,
          },
          duration: 0,
          frequency: 0,
          time_between_alerts: 0,
          destination: "dest1",
          is_real_time: true,
          stream_type: "logs",
        },
        {
          name: "alert2",
          stream: "default",
          query: {
            sql: "select * from  'default'",
            from: 0,
            size: 100,
            start_time: 0,
            end_time: 0,
            query_type: "",
            track_total_hits: false,
          },
          condition: {
            column: "code",
            operator: "EqualTo",
            ignore_case: null,
            value: 500,
          },
          duration: 1,
          frequency: 2,
          time_between_alerts: 2,
          destination: "dest1",
          is_real_time: false,
          stream_type: "logs",
        },
      ],
    },
  },
  destinations: {
    get: [
      {
        name: "dest1",
        url: "abc",
        method: "post",
        headers: {},
        template: {
          name: "template1",
          body: '"This is string."',
          isDefault: true,
        },
      },
      {
        name: "dest2",
        url: "https://join.slack.com/share/enQtNTAyMDAzNDY",
        method: "post",
        headers: {},
        template: {
          name: "Template3",
          body: '\r\n[\r\n  {\r\n    "labels": {\r\n        "alertname": "{alert_name}",\r\n        "stream": "{stream_name}",\r\n        "organization": "{org_name}",\r\n        "alerttype": "{alert_type}",\r\n        "severity": "critical"\r\n    },\r\n    "annotations": {\r\n        "timestamp": "{timestamp}"\r\n    }\r\n  }\r\n]',
          isDefault: true,
        },
      },
    ],
  },
  templates: {
    get: [
      {
        name: "Template2",
        body: '\r\n{\r\n  "text": "{alert_name} is active"\r\n}',
        isDefault: true,
      },
      {
        name: "Template3",
        body: '\r\n[\r\n  {\r\n    "labels": {\r\n        "alertname": "{alert_name}",\r\n        "stream": "{stream_name}",\r\n        "organization": "{org_name}",\r\n        "alerttype": "{alert_type}",\r\n        "severity": "critical"\r\n    },\r\n    "annotations": {\r\n        "timestamp": "{timestamp}"\r\n    }\r\n  }\r\n]',
        isDefault: true,
      },
      {
        name: "template1",
        body: '"This is string."',
        isDefault: true,
      },
    ],
  },
};
